File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.py
import sys

input = sys.stdin.readline

INF = float("inf")


def calc():
    result = 0
    for i in range(n - 1):
        result += abs(selected[i] - selected[i + 1])
    return result


def permutation(curr):
    global answer
    if curr == n:
        answer = max(answer, calc())
        return

    for i in range(n):
        if not is_selected[i]:
            is_selected[i] = True
            selected[curr] = numbers[i]
            permutation(curr + 1)
            is_selected[i] = False


n = int(input())
numbers = list(map(int, input().split()))

selected = [0] * n
is_selected = [False] * n

answer = -INF

permutation(0)

print(answer)

File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int n;
    static int[] array;
    static boolean[] isSelected;
    static int answer = Integer.MIN_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        array = new int[n];
        isSelected = new boolean[n];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            array[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 0; i < n; i++) {
            isSelected[i] = true;
            solve(1, array[i], 0);
            isSelected[i] = false;
        }

        System.out.println(answer);

    }

    private static void solve(int curr, int prev, int sum) {
        if (curr == n) {
            answer = Math.max(answer, sum);
            return;
        }

        for (int i = 0; i < n; i++) {
            if (!isSelected[i]) {
                isSelected[i] = true;
                solve(curr + 1, array[i], sum + Math.abs(prev - array[i]));
                isSelected[i] = false;
            }
        }
    }
}